// express-request-router/express-request-router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const express = require("express");
const request = require("request-promise-native");
const bodyParser = require("body-parser");
const router = express.Router();
;
router.use(bodyParser.json());
router.post('/', async (req, res, next) => {
    let requestOptions = req.body;
    let response;
    try {
        response = await request({
            url: requestOptions.url,
            method: requestOptions.method || 'get',
            headers: requestOptions.headers || {},
            body: requestOptions.body || '',
            followRedirect: requestOptions.redirect === 'follow',
            resolveWithFullResponse: true,
            proxy: requestOptions.proxy || null
        });
    }
    catch (error) {
        if (error.name === 'StatusCodeError') {
            res.status(error.statusCode);
            response = error.response;
        }
        else {
            throw error;
        }
    }
    finally {
        if ('headers' in response)
            for (let header in response.headers)
                res.setHeader(header, response.headers[header]);
        res.end(response.body);
    }
});
exports.default = router;
